waspc/src/Analyzer/Evaluator/TH.hs

```diff
@@ -0,0 +1,246 @@
+-- always changed to lowercase).
+--
+-- __Example__
+--
+-- @
+-- {-# LANGUAGE TemplateHaskell #-}
+-- data Person = Person { name :: String, age :: Int }
+-- makeDecl ''Person
+-- -- "IsDeclType Person" instance is generated
+-- @
+makeDecl :: Name -> Q [Dec]
+makeDecl typeName = do
+  (TyConI tyCon) <- reify typeName
+  (tyConName, con) <- case tyCon of
+    (DataD _ name [] _ [con] _) -> pure (name, con)
+    (NewtypeD _ name [] _ con _) -> pure (name, con)
+    _ -> fail "Invalid name for makeDecl"
+  let instanceType = conT ''IsDeclType `appT` conT tyConName
+  instanceDecs <- genDecl con
+  sequence [instanceD (return []) instanceType instanceDecs]
+
+-- | @makeEnum ''Type@ writes an @IsEnumType@ instance for @Type@. A type
+-- error is raised if @Type@ does not fit the criteria described in the
+-- definition of @IsEnumType@.
+--
+-- In addition to satisfying the requirements of "IsEnumType", the generaed
+-- instance for @Type@ has @enumTypeName == "Type"@ (the name is not modified
+-- at all).
+--
+-- __Example__
+--
+-- @
+-- {-# LANGUAGE TemplateHaskell #-}
+-- data Job = Programmer | Manager
+-- makeEnum ''Job
+-- -- "IsEnumType Job" instance is generated
+-- @
+makeEnum :: Name -> Q [Dec]
+makeEnum typeName = do
+  (TyConI tyCon) <- reify typeName
+  (tyConName, cons) <- case tyCon of
+    (DataD _ name [] _ cons _) -> pure (name, cons)
+    (NewtypeD _ name [] _ con _) -> pure (name, [con])
+    _ -> fail "Invalid name for makeEnum"
+  let instanceType = conT ''IsEnumType `appT` conT tyConName
+  conNames <- enumConNames cons
+  instanceDecs <- genEnum tyConName conNames
+  sequence [instanceD (return []) instanceType instanceDecs]
+
+-- ========================================
+-- IsDeclType generation
+-- ========================================
+
+-- | Top-level "IsDeclType" instance generator.
+genDecl :: Con -> Q [DecQ]
+-- The constructor is in the form @data Type = Type x@
+genDecl (NormalC name [(_, typ)]) = genPrimDecl name typ
+-- The constructor is in the form @data Type = Type x1 x2 ... xn@, which is not valid for a decl
+genDecl (NormalC name _) = fail $ "Too many non-record values in makeDecl for " ++ show name
+-- The constructor is in the form @data Type = Type { k1 :: f1, ..., kn :: fn }
+genDecl (RecC name recs) = genRecDecl name $ map (\(recName, _, typ) -> (recName, typ)) recs
+-- The constructor is in an unsupported form
+genDecl _ = fail "makeDecl on non-decl type"
+
+-- | Create an "IsDeclType" instance for types in the form @data Type = Type x@
+genPrimDecl :: Name -> Type -> Q [DecQ]
+genPrimDecl name typ =
+  pure
+    [ func 'declTypeName $ lowerNameStrE name,
+      func 'declTypeBodyType $ genTypeE typ,
+      func 'declTypeFromAST [|build $ $(conE name) <$> $(genEvaluatorE typ)|]
+    ]
+
+-- | For decls with record constructors, i.e. @data Fields = Fields { a :: String, b :: String }
+genRecDecl :: Name -> [(Name, Type)] -> Q [DecQ]
```

Comment:
I would go with singular, `Record`, for sure. I also feel that `DataConstructor` part tells us what it operates on, and since the types don't tell us that, I would rather keep it in the name.

Reply:
Renamed.